File: solutions/zig/resistor-color-trio/2/resistor_color_trio.zig
const std = @import("std");
const mem = std.mem;

pub const ColorBand = enum(usize) {
    black = 0,
    brown = 1,
    red = 2,
    orange = 3,
    yellow = 4,
    green = 5,
    blue = 6,
    violet = 7,
    grey = 8,
    white = 9,
};

const Unit = struct {
    threshold: u128,
    divisor: f64,
    prefix: []const u8,
};

const units = [_]Unit{
    .{ .threshold = 1_000_000_000, .divisor = 1_000_000_000.0, .prefix = "giga" },
    .{ .threshold = 1_000_000, .divisor = 1_000_000.0, .prefix = "mega" },
    .{ .threshold = 1_000, .divisor = 1_000.0, .prefix = "kilo" },
    .{ .threshold = 0, .divisor = 1.0, .prefix = "" },
};

pub fn label(allocator: mem.Allocator, colors: []const ColorBand) mem.Allocator.Error![]u8 {
    const base: u128 = @intFromEnum(colors[0]) * 10 + @intFromEnum(colors[1]);
    const exp = @intFromEnum(colors[2]);
    const ohms: u128 = base * std.math.pow(u128, 10, exp);

    var f_ohms: f64 = @floatFromInt(ohms);
    var prefix: []const u8 = "";

    for (units) |unit| {
        if (ohms >= unit.threshold) {
            f_ohms /= unit.divisor;
            prefix = unit.prefix;
            break;
        }
    }

    return std.fmt.allocPrint(allocator, "{d} {s}ohms", .{f_ohms, prefix});
}

File: solutions/elixir/resistor-color-trio/2/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do

  @colors %{
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9
  }

  defp get_unit(value) when value >= 1_000_000_000, do: %{divisor: 1_000_000_000.0, prefix: :gigaohms}
  defp get_unit(value) when value >= 1_000_000, do: %{divisor: 1_000_000.0, prefix: :megaohms}
  defp get_unit(value) when value >= 1_000, do: %{divisor: 1_000.0, prefix: :kiloohms}
  defp get_unit(_value), do: %{divisor: 1.0, prefix: :ohms}

  defp value(colors) do
    colors
    |> Enum.take(2)
    |> Enum.map(&Map.get(@colors, &1))
    |> Enum.join()
    |> String.to_integer()
    |> Kernel.*(10 ** Map.get(@colors, Enum.at(colors, 2)))
  end

  @doc """
  Calculate the resistance value in ohms from resistor colors
  """
  
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms | :megaohms | :gigaohms}
  def label(colors) do
    colors 
    |> value() 
    |> then(fn resistance ->
      unit = get_unit(resistance)
      {resistance / unit.divisor, unit.prefix}
    end)
  end
end
